crates/bifrost/src/providers/replicated_loglet/tasks/find_tail.rs

```diff
@@ -0,0 +1,645 @@
+                                return FindTailResult::Sealed {
+                                    global_tail: max_local_tail,
+                                };
+                            } else {
+                                // F-majority sealed, but tail needs repair in range
+                                // [current_known_tail..max_local_tail]
+                                //
+                                // todo: Although we have f-majority, it's not always guaranteed that we are
+                                // able to form a write-quorum within this set of nodes. For instance, if
+                                // replication-factor is 4 in a nodeset of 5 nodes, F-majority is 2 nodes which
+                                // isn't enough to form write-quorum. In this case, we need to wait for more
+                                // nodes to chime in. But we _only_ need this if max_local_tail is higher than
+                                // current_known_global. Once we have enough sealed node that match
+                                // write-quorum **and** f-majority, then we can repair the tail.
+                                if nodeset_checker.check_write_quorum(NodeTailStatus::is_known) {
+                                    // We can repair.
+                                    todo!("Tail repair is not implemented yet")
+                                } else {
+                                    // wait for more nodes
+                                    break 'check_nodeset;
+                                }
+                            }
+                        };
+
+                        // # Some Sealed
+                        // Not f-majority sealed, but is there any node that's sealed?
+                        // Run a `SealTask` to assist in sealing.
+                        if nodeset_checker.any(NodeTailStatus::is_known_sealed) {
+                            // run seal task then retry the find-tail check.
+                            let seal_task = SealTask::new(
+                                self.task_center.clone(),
+                                self.my_params.clone(),
+                                self.logservers_rpc.seal.clone(),
+                                self.known_global_tail.clone(),
+                            );
+                            debug!("Detected unsealed nodes. Running seal task to assist in sealing loglet_id={}", self.my_params.loglet_id);
+                            // This returns when we have f-majority sealed.
+                            if let Err(e) = seal_task.run(self.networking.clone()).await {
+                                return FindTailResult::Error(format!(
+                                    "Failed to seal loglet_id={}: {:?}",
+                                    self.my_params.loglet_id, e
+                                ));
+                            }
+                            inflight_info_requests.abort_all();
```

Comment:
Should this be moved before the `seal_task.run()` ?

Reply:
It'd be confusing to have two places where tasks are aborted. In general, as I mentioned in the PR description, this code is very verbose (intentionally) and it'll be refactored at a later stage once the algorithm itself is tested and its correctness is verified. So I wouldn't worry so much about the cosmetics.